### DISCUSSION TITLE
Form() does not treat emptry string as None anymore

### DISCUSSION THREAD

**Discussion by fnep** [fastapi/fastapi#12377] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Form
from typing import Annotated

from fastapi import FastAPI

app = FastAPI()


@app.post("/")
def read_root(
    user_id: Annotated[int | None, Form()] = None,
):
    return user_id
```

### Description

In earlier versions, FastAPI treated an empty `Form()` value (like it is given from an HTML `<form>` → `<input>` element) as `None`. Since about v0.113.0 this is treated as an empty string, what breaks many of my forms.

In `fastapi==0.115.0`:
```bash
❯ curl  --form user_id= http://127.0.0.1:8000/
{"detail":[{"type":"int_parsing","loc":["body","user_id"],"msg":"Input should be a valid integer, unable to parse string as an integer","input":""}]}⏎                                                                        
```

whereas in `fastapi==0.112.2`:
```bash
❯ curl  --form user_id= http://127.0.0.1:8000/
null
```

### Operating System

Linux, macOS

### FastAPI Version

0.115.0

### Pydantic Version

2.9.0

### Python Version

3.12.1

### Additional Context

I'm not entirely sure if this is a bug, but it is at least a regression. 

Working around it is a big problem, also because finding those places is hard. Form values usually come from a browser. There is no way to set this value to None other than omitting it, what is hard with plain HTML.

**Comment by sinisaos** [comment]
@fnep As I tried to explain in [this discussion](https://github.com/fastapi/fastapi/discussions/12332#discussioncomment-10805006), I think the main problem is that FastAPI treats an `empty` string as a `None` value, but I don't think that's true. As I suggested in that discussion (I'm deeply sorry if I'm wrong), but if you patch your FastAPI code with an `empty string` check along with the `None` check, everything works. Replace this [line](https://github.com/fastapi/fastapi/blob/847296e885ed83bc333b6cc0a3000d6242083b87/fastapi/dependencies/utils.py#L691) with this
```python
if value is None or not value: # check empty string along None
```
Result is:

```bash
> curl --form user_id= http://127.0.0.1:8000/
null
```

**Reply by fnep** [reply]
@sinisaos I get the point and understand your view. Only, how am I supposed to ever give a real `None` value via `Form()`? I can only omit it. I think we cannot treat values to `Form()` the same as let's say values to `Body()`. It makes handling normal Browser behavior cumbersome.

Anyhow, for me, the most important part would be to have an “official” statement, best in the docs, that this new behavior will remain and is wanted. I will most likely be able to update my code, but it is a lot of work.

**Reply by sinisaos** [reply]
> Only, how am I supposed to ever give a real `None` value via `Form()`? I can only omit it. I think we cannot treat values to `Form()` the same as let's say values to `Body()`. It makes handling normal Browser behavior cumbersome.

Maybe I didn't explain well (sorry) or we didn't understand each other. You're not sending a `None` value but an `empty string`(`""`) from the browser. Now the validation fails because FastAPI checks for a `None` value, not an `empty string` (`"" != None`). By changing the line of code from the previous comment, we check for an `empty string` along with the `None` check. 

```python
if value is None or not value: # check empty string along None
# or 
if value is None or value == "": # check empty string along None
```

Can you try that? In my case with that change validation pass.

> Anyhow, for me, the most important part would be to have an “official” statement, best in the docs, that this new behavior will remain and is wanted. I will most likely be able to update my code, but it is a lot of work.

For that you need to wait for the someone relevant from the FastAPI team. Cheers.

**Reply by fnep** [reply]
Changing the line would definitely fix the issue, but something like:
```python3
if value is None or (isinstance(field.field_info, Form) and value == ""):
```
... would suit better. 

I think I understood the case and am aware that the Browser is sending an empty string. My point is more that `Form()` is made for browser `<form>` elements, and they behave different from a `curl` or `fetch()` for example. This makes it hard to deal with. 

Even if I prefer the old behavior, I can understand this change.

Only some confirmation would be great. :) Maybe @tiangolo could take a glance.

**Reply by sinisaos** [reply]
> Changing the line would definitely fix the issue, but something like:
> 
> ```python
> if value is None or (isinstance(field.field_info, Form) and value == ""):
> ```
> 
> ... would suit better.

That's even better  :thumbsup: .

**Comment by YuriiMotov** [comment]
Duplicate of https://github.com/fastapi/fastapi/discussions/12150
Let's close this topic and track it there

**Comment by YuriiMotov** [comment, marked as answer]
This was fixed by https://github.com/fastapi/fastapi/pull/13537 and available in FastAPI [0.123.2](https://github.com/fastapi/fastapi/releases/tag/0.123.2)
